### ISSUE TITLE
Add Task to warmup caches

### ISSUE THREAD

**Issue by mrimann** [TYPO3/Surf#427]
## Problem
We deploy Flow applications and Neos CMS instances via Gitlab CI with Surf since years. Recently, we noticed that some things broke due to a possible race condition. Usually our Gitlab Pipeline looks like this:

Step 1: Deploy via Surf

Step 2a: Run some re-indexing task (e.g. build up ElasticSearch indexes)
Step 2b: Trigger a run of our version monitoring tool (which in turn sends a GET request to an URL of the freshly deployed application)

Steps 2a and 2b are executed in parallel. And also in parallel there could be some real-world traffic (2c) happening.

Depending on the timing, it could happen that the application is in a "caches are being built, please wait" state when 2a, 2b or 2c are happening. For 2c (humans) that's OK - they'll wait or try again later. But 2a an 2b will fail (and, depending on the config, mark the deployment step as failed).

## Tested things to avoid the problem

We *could* set up separate vHosts for smoke-testing based on the `next` release. But since we're talking 80+ installations, that would be a pretty substantial config overhead.

We *could* make the steps 2a and 2b happen sequentially instead of parallelized. That would probably solve the issue for 90% cases. For installations without real-world traffic this number might be closer to 100% - but for those with traffic from the outside (2c) the problem could still happen.

We *could* make our deployment steps for 2a and 2b more "robust" by adding some delay and/or repetitions to make sure a temporary "locked, caches are built" situation, to be less of a problem. Doable, but not solving the issue on the root.

We *could* extend the application's deployment steps to contain an additional `Neos/Flow/RunCommandTask` or even a `LocalShellTask`. That would need to be integrated into dozens of Surf config files - doable, but a bit of work.

## Proposed solution
If we could warm-up the caches in the first step, within the deployment of the application and **before** switching the deployed version to be the `current`-release, that should solve the problem before it can happen at all.

Basically just a `FLOW_CONTEXT=Production path/to/releases/next/flow flow:cache:warmup` is needed.

And if that could become either a default task to always happen for Flow- and Neos-Instances, that could be beneficial for everyone I think. So far I don't see anything that would speak against doing so.

## Specifications

  - Surf Version: 2.0.x
  - Application: Neos CMS or Flow based applications
  - PHP Version: 7.4.x
  - Platform: Ubuntu 20.04.x
  - Environment (CI): Gitlab CI

**Comment by simonschaufi** [comment]
Sounds valid. Will you create a PR?

**Comment by mrimann** [comment]
I've asked in #guild-deployment channel on Neos-Slack and hope for some feedback on how to go on with this one, regarding the order of Tasks and what to do in a rollback scenario.

**Comment by kaystrobach** [comment]
The solution is fine for file based caches. But will fail for redis or database caches if they are not bound to the specific release

**Comment by t3easy** [comment]
@kaystrobach we already discussed that. The rollback have to flush and warmup the caches on the old release. Then db caches should be fine.

**Comment by kaystrobach** [comment]
Right - just to ensure, you get what i meant:

Warming up the cache on the „next“ release can already break the „prod“ env if prod and next use the same cache (e.g. redis) but this is imho in the scope of the developer ...

**Comment by t3easy** [comment]
Yes. The "before switch" is the hardest part. Or is it possible to prefix cache tables/entrees somehow with Neos?

**Comment by kaystrobach** [comment]
If cache is stored in files (default) all is fine - no Problem warmed up isolated. If you use redis or memcache or similar technologies you just need to follow the rules to not have BCs from deployment to deployment. E.g. add columns, migrate date and write new and old, and remove columns with a new deployment.

But this is imho up to the developer team. 

We can solve that for standard cases, but not for all cases.

**Comment by mrimann** [comment]
Thanks for all the input @t3easy and @kaystrobach and Johannes in the Slack channel! I think I've now got the following image on the whole situation:
- it does not affect the per deployment/release file cache (nothing new here)
- it affects DB / Redis caches that are not isolated from deployment to deployment (e.g. shared between releases) in general, but we need to *differentiate the type of caches*:
-- certain caches like e.g. Session-Caches per User are most likely not sensible to this "problem" and keeping them as is shouldn't be an issue at all, even across releases
-- other caches like e.b. Routing-Config dependant Routing-Caches that could get modified by the new version being rolled out are more sensible to this topic (and warming up caches for `next` could break `current` in the meantime.

And I agree full with Kay's thougt: It depends, from project to project.

So my proposal would be to
- add the task as optional (e.g. feature-flagged, default off)
- add a config option to enable this task for Flow/Neos deployments
- let the Task happen right before switching the symlink to the new deployment if it's enabled
- add documentation mentioning this new optional task and that it needs to be activated when wanted - with a sidenote on possible problems this could bring with itself.

@simonschaufi OK for you?

**Comment by simonschaufi** [comment]
@mrimann I could even simply it even more and just provide the task and you can attach it to whatever stage you want in your own deployment config. Alternatively I could add it after symlink switch and you can remove it and attach it again before the symlink switch. What do you prefer?

**Comment by mrimann** [comment]
We (@simonschaufi) and I discussed that today and will go on with splitting the second part of my last proposal into *two config options*:

- if one does not explicitely enable cache-warmup tasks (default = off), nothing would change, the new WarmupTask will never be executed
- if one enables e.g. `enableCacheWarmupBeforeSwitchingToNewRelease`, the Cache-Warmup task hooked in the deployment order as last task **before** switching symlinks to the new deployment
- if one enables e.g. `enableCacheWarmupAfterSwitchingToNewRelease`, the same happens, but just **after** switching to the new deployment

I think this is easier for the users, if they can just enable this by enabling one of the newly added feature-flags / config options.

**Comment by simonschaufi** [comment]
@mrimann can you do some tests and give me feedback if it works like expected?

**Comment by mrimann** [comment]
I just added test-feedback to the pull-request: https://github.com/TYPO3/Surf/pull/429
